compiler/rustc_trait_selection/src/solve/alias_relate.rs

```diff
@@ -84,3 +93,105 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
         }
     }
 }
+
+enum IgnoreAliases {
+    Yes,
+    No,
+}
+
+impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
+    /// In case a rigid term refers to a placeholder which is not referenced by the
+    /// alias, the alias cannot be normalized to that rigid term unless it contains
+    /// either inference variables or these placeholders are referenced in a term
+    /// of a `Projection`-clause in the environment.
+    fn alias_cannot_name_placeholder_in_rigid(
+        &mut self,
+        param_env: ty::ParamEnv<'tcx>,
+        rigid_term: ty::Term<'tcx>,
+        alias: ty::Term<'tcx>,
+    ) -> bool {
+        // Check that the rigid term is actually rigid.
+        if rigid_term.to_alias_term().is_some() || alias.to_alias_term().is_none() {
+            return false;
+        }
+
+        // If the alias has any type or const inference variables,
+        // do not try to apply the fast path as these inference variables
+        // may resolve to something containing placeholders.
+        if alias.has_non_region_infer() {
+            return false;
+        }
+
+        let mut referenced_placeholders =
+            self.collect_placeholders_in_term(rigid_term, IgnoreAliases::Yes);
+        for clause in param_env.caller_bounds() {
+            match clause.kind().skip_binder() {
+                ty::ClauseKind::Projection(ty::ProjectionPredicate { term, .. }) => {
+                    if term.has_non_region_infer() {
+                        return false;
+                    }
+
+                    let env_term_placeholders =
+                        self.collect_placeholders_in_term(term, IgnoreAliases::No);
+                    #[allow(rustc::potential_query_instability)]
+                    referenced_placeholders.retain(|p| !env_term_placeholders.contains(p));
+                }
+                ty::ClauseKind::Trait(_)
+                | ty::ClauseKind::TypeOutlives(_)
+                | ty::ClauseKind::RegionOutlives(_)
+                | ty::ClauseKind::ConstArgHasType(..)
+                | ty::ClauseKind::WellFormed(_)
+                | ty::ClauseKind::ConstEvaluatable(_) => continue,
+            }
+        }
+
+        if referenced_placeholders.is_empty() {
+            return false;
+        }
+
+        let alias_placeholders = self.collect_placeholders_in_term(alias, IgnoreAliases::No);
+        // If the rigid term references a placeholder not mentioned by the alias,
```

Comment:
```suggestion
        // If the rigid term references a placeholder not mentioned by the alias, and isn't potentially reachable by a projection-predicate,
```

or something, to explain the way that the projection predicate search above influences this search